From 683336267051a0d2689aea1a1075828b5901ddfd Mon Sep 17 00:00:00 2001
From: Huacai Chen <chenhc@lemote.com>
Date: Fri, 4 Nov 2016 14:41:50 +0800
Subject: [PATCH 114/130] R8168: Detect and recover weird rx hang bug (V2)

V2 of this patch is in order to avoid kernel panic.

Change-Id: I53ddd0ffcbe3c427ad56e90979c7d62d6a5751c4
Signed-off-by: Huacai Chen <chenhc@lemote.com>
---
 drivers/net/ethernet/realtek/r8168/r8168_n.c | 12 +++++-------
 1 file changed, 5 insertions(+), 7 deletions(-)

diff --git a/drivers/net/ethernet/realtek/r8168/r8168_n.c b/drivers/net/ethernet/realtek/r8168/r8168_n.c
index 644375b1..3cac4b94 100644
--- a/drivers/net/ethernet/realtek/r8168/r8168_n.c
+++ b/drivers/net/ethernet/realtek/r8168/r8168_n.c
@@ -24568,7 +24568,7 @@ static void rtl8168_recover(struct rtl8168_private *tp)
         struct net_device *dev = tp->dev;
 
         spin_lock_irqsave(&tp->lock, flags);
-
+#if 0
         for (i = 0; i < NUM_RX_DESC; i++) {
                 if (tp->Rx_skbuff[i])
                         rtl8168_free_rx_skb(tp, tp->Rx_skbuff + i,
@@ -24587,7 +24587,7 @@ static void rtl8168_recover(struct rtl8168_private *tp)
         }
 
         rtl8168_mark_as_last_descriptor(tp->RxDescArray + NUM_RX_DESC - 1);
-
+#endif
         rtl8168_set_speed(dev, tp->autoneg, tp->speed, tp->duplex);
 
         spin_unlock_irqrestore(&tp->lock, flags);
@@ -25239,9 +25239,6 @@ rtl8168_rx_interrupt(struct net_device *dev,
         rx_left = NUM_RX_DESC + tp->dirty_rx - cur_rx;
         rx_left = rtl8168_rx_quota(rx_left, (u32)rx_quota);
 
-	if (intsts & RxOK)
-		tp->old_rx = tp->cur_rx;
-
         for (; rx_left > 0; rx_left--) {
                 rmb();
                 status = le32_to_cpu(desc->opts1);
@@ -25348,9 +25345,10 @@ process_pkt:
                 printk(KERN_EMERG "%s: Rx buffers exhausted\n", dev->name);
 
 	if (intsts & RX_EVENT) {
-		if (tp->cur_rx != tp->old_rx)
+		if (tp->cur_rx != tp->old_rx) {
+                        tp->old_rx = tp->cur_rx;
                         tp->weird_hang_recheck = 0;
-		else {
+		} else {
                         if (tp->weird_hang_recheck < CHKCNT)
                                 tp->weird_hang_recheck += 1;
                         else {
-- 
2.11.0

